README.md
# homebridge-solaredge-inverter
A [SolarEdge](https://www.solaredge.com) Inverter plugin for
[Homebridge](https://github.com/nfarina/homebridge).  This creates a a set of Light Sensors in homekit,
 where the LUX reading is actually the  power generation in KiloWatts.  There is also an option to display battery level if applicable.

This code is heavily based on the work of Stog's [homebridge-fronius-inverter](https://github.com/Stog/homebridge-fronius-inverter) accessory.

# Installation
Run these commands:

    % sudo npm install -g homebridge
    % sudo npm install -g homebridge-solaredge-inverter


NB: If you install homebridge like this:

    sudo npm install -g --unsafe-perm homebridge

Then all subsequent installations must be like this:

    sudo npm install -g --unsafe-perm homebridge-solaredge-inverter

# Configuration

Example accessory config (needs to be added to the homebridge config.json):
 ...

		"accessories": [
			{
				"name": "SolarEdge Inverter",
				"manufacturer": "SolarEdge",
				"model": "SE10000H-US000BNU4",
				"serial": "myserialno",
				"site_id": "mysiteid",
				"api_key": "longapikey",
				"update_interval": 15,
				"accessory": "SolarEdge Inverter",
				"current": true,
				"currentWatts": true,
				"last_day": false,
				"last_month": false,
				"last_year": false,
				"life_time": true
				"battery": true
				"debug": false
			}
		]
 ...

### Config Explanation:

Field           						| Description
----------------------------|------------
**accessory**   						| (required) Must always be "SolarEdge Inverter".
**name**										| (required) The name you want to use for for the power level widget.
**site_id**  								| (required) The Site ID for your SolarEdge installation.
**api_key**		  						| (required) The API Key for the administration of your SolarEdge site.
**manufacturer**						| (optional) This shows up in the homekit accessory Characteristics.
**model**										| (optional) This shows up in the homekit accessory Characteristics.
**serial**									| (optional) This shows up in the homekit accessory Characteristics.
**update_interval**					| (optional) The frequency to poll the SolarEdge API in minutes (defaults to 15).
**debug**										| (optional) Enables additional logging.
**current**									| (required) Display current power (kW).
**currentWatts**						| (optional) Display current power in W vs. kW.
**last_day**								| (required) Display Last Day power (kW).
**last_month**							| (required) Display Last Month power (kW).
**last_year**								| (required) Display Last Year power (kW).
**life_time**								| (required) Display Life Time power (kW).
**battery**									| (required) Display Battery Level (%).



CHANGELOG.md
# Changelog


## 0.2.7 (2021-11-09)

#### Changes

- Changed verbiage for Current Month & Current Year (Issue #27).

## 0.2.6 (2021-10-05)

#### Changes

- Updated dependencies.

## 0.2.5 (2021-03-08)

#### Changes

- Utilizes timed updates based on update_interval (in minutes).
- Adjusted power responses of 0 to return 0.0001 Lumens to Homekit (Issue #19).

## 0.2.4 (2021-01-02)

#### Changes

- Allows for multiple instances to be loaded for the same Inverter with different measurements - Lifetime vs Current Power, without forcing a Current Power measurement on each instance (Issue #16).

## 0.2.3 (2021-01-02)

#### Changes

- Fixed bug at end of day, was not updating current power when current power is 0 (Issue #14, Issue #15).
- Fixed a README.md typo.

## 0.2.2 (2020-12-31)

#### Changes

- Entire re-write to a push vs. pull update model, minimizing SolarEdge API calls (Issue #11).
- Updated math rounding.
- Updated README.md.
- Good-bye 2020!

## 0.2.1 (2020-12-29)

#### Changes

- Bug fixed.  Was not displaying anything but current power.

## 0.2.0 (2020-12-23)

#### Changes

- New config.schema.json -- REQUIRES changes to the config to remove display section
- Added option to display Current Power in Watts vs. kW

## 0.1.7 (2020-12-14)

#### Changes

- Fixed debug logging

## 0.1.6 (2020-12-14)

#### Changes

- Fixed some fat-fingering :P

## 0.1.5 (2020-12-14)

#### Changes

- Modified config.schema.json
- Added update_interval to allow the use of cached data vs. API polling
- Added a debug option for logging (the default is silent)

## 0.1.4 (2020-11-09)

#### Changes

- Modified config.schema.json
- Added backwards compatibility if a user neglected to add *display* config section

## 0.1.3 (2020-11-03)

#### Changes

- Fixed a typo in config.schema.json

## 0.1.2 (2020-11-03)

#### Changes

- Added an optional BatteryLevel display

## 0.1.1 (2020-11-01)

#### Changes

- Modified the math to work in absolute values (my homebridge instance needed this)

## 0.1.0 (2020-11-01)

#### Changes contributed by tomrush

- Added additional LightSensor's for:
        Current Day
        Last Month
        Last Year
        Life Time
- Converted units to kW and limited at 2 decimal places.
- Updated Config schema to allow for new settings.
- Updated documentation
- Added Git ignore file

## 0.0.4 (2020-04-28)

#### Changes

- Cleaned up comments in index.js.
- Updated Changelog & Readme.

## 0.0.3 (2020-04-28)

#### Changes

- Added npm publish workflow.
- Updated packages.json.

## 0.0.2 (2020-04-27)

#### Changes

- Bug fix.
- Added information and sample config in Readme.
- Added shout-out in Readme.

## 0.0.1 (2020-04-27)

#### Initial Commit


index.js
const axios = require('axios');
const setupCache = require('axios-cache-adapter').setupCache;

var Service, Characteristic;

const DEF_MIN_LUX = 0,
      DEF_MAX_LUX = 65535;

const DISPLAY_USAGE_SENSORS = 0;

const PLUGIN_NAME   = 'homebridge-solaredge-inverter';
const ACCESSORY_NAME = 'SolarEdge Inverter';

module.exports = function(homebridge) {
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    homebridge.registerAccessory(PLUGIN_NAME, ACCESSORY_NAME, SolarEdgeInverter);
}

/**
 * Setup Cache For Axios to prevent additional requests
 */
const cache = setupCache({
	maxAge: 1 * 60 * 1000, //in ms
	readHeaders: false,
	// For this example to work we disable query exclusion
	exclude: { query: false }
})

const api = axios.create({
	adapter: cache.adapter
})

function isEmptyObject(obj) {
  for (var key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      return false;
    }
  }
  return true;
}

/**
 * Main API request with site overview data
 *
 * @param {siteID} the SolarEdge Site ID to be queried
 * @param {apiKey} the SolarEdge monitoring API Key for access to the Site
 */
const getInverterData = async(siteID, apiKey, update_interval) => {
	try {
	    return await api.get('https://monitoringapi.solaredge.com/site/'+siteID+'/overview?api_key='+apiKey, {
	    	cache: {
	    	maxAge: update_interval
	    	}
	    })
	} catch (error) {
	    console.error(error)
	}
}

/**
 * Gets and returns the accessory's value in the correct format.
 *
 * @param {siteID} the SolarEdge Site ID to be queried
 * @param {apiKey} the SolarEdge monitoring API Key for access to the Site
 * @param (log) access to the homebridge logfile
 * @return {bool} the value for the accessory
 */
const getAccessoryValue = async(that) => {
	if (that.debug) {
		that.log.info('Calling API');
	}
	// To Do: Need to handle if no connection
	const inverterData = await getInverterData(that.siteID, that.apiKey, that.update_interval)

	if(inverterData) {
		if (that.debug) {
			if (inverterData.request.fromCache !== true) {
				that.log.info('Data from API', inverterData.data.overview);
			} else {
				that.log.info('Data from cache', inverterData.data.overview);
			}
		}
		if(inverterData.data.overview) {
			return inverterData.data.overview;
		}
		else {
			return null
		}
	} else {
		return null
	}
}

/**
 * API request with power flow data
 *
 * @param {siteID} the SolarEdge Site ID to be queried
 * @param {apiKey} the SolarEdge monitoring API Key for access to the Site
 */
const getPowerFlowData = async(siteID, apiKey, update_interval) => {
	try {
	    return await api.get('https://monitoringapi.solaredge.com/site/'+siteID+'/currentPowerFlow?api_key='+apiKey, {
	    	cache: {
	    	maxAge: update_interval
	    	}
	    })
	} catch (error) {
	    console.error(error)
	}
}

/**
 * Gets and returns the battery's charge value in the correct format.
 */
const getBatteryValues = async (that) => {
	if (that.debug) {
		that.log.info('Calling Flow API');
	}
	// To Do: Need to handle if no connection
	const powerFlowData = await getPowerFlowData(that.siteID, that.apiKey, that.update_interval)

	if(powerFlowData) {
		if (that.debug) {
			if (powerFlowData.request.fromCache !== true) {
				that.log.info('Data from Power Flow API', powerFlowData.data.siteCurrentPowerFlow);
			} else {
			that.log.info('Data from Power Flow cache', powerFlowData.data.siteCurrentPowerFlow);
			}
		}
		if(powerFlowData.data.siteCurrentPowerFlow) {
			return powerFlowData.data.siteCurrentPowerFlow;
		}
		else {
			return null
		}
	} else {
		return null
	}
}

const update = async(that) => {
	if(that.currentPower || that.lastDayPower || that.lastMonth || that.lastYear || that.lifeTime) {
		const accessoryValue = await getAccessoryValue(that);
		let power = 0;
		if (accessoryValue) {
			if(that.currentPower) {
				if(parseFloat(accessoryValue.currentPower.power) > 0) {
					if(that.currentWatts) {
						power = Math.abs(Math.round((accessoryValue.currentPower.power + Number.EPSILON) *10) /10)
					} else {
						power = Math.abs(Math.round(((accessoryValue.currentPower.power / 1000) + Number.EPSILON) *10) /10)
					}
				}
				that.currentPower
					.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
					.updateValue(power ? power : 0.0001)
			}
			if(that.lastDayPower) {
				power = Math.abs(Math.round(((accessoryValue.lastDayData.energy / 1000) + Number.EPSILON) *10) /10)
				that.lastDayPower
					.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
					.updateValue(power ? power : 0.0001)
			}
			if(that.lastMonth) {
				power = Math.abs(Math.round(((accessoryValue.lastMonthData.energy / 1000) + Number.EPSILON) *10) /10)
				that.lastMonth
					.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
					.updateValue(power ? power : 0.0001)
			}
			if(that.lastYear) {
				power = Math.abs(Math.round(((accessoryValue.lastYearData.energy / 1000) + Number.EPSILON) *10) /10)
				that.lastYear
					.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
					.updateValue(power ? power : 0.0001)
			}
			if(that.lifeTime) {
				power = Math.abs(Math.round(((accessoryValue.lifeTimeData.energy / 1000) + Number.EPSILON) *10) /10)
				that.lifeTime
					.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
					.updateValue(power ? power : 0.0001)
			}
		}
	}
	if(that.battery){
		const batteryValues = await getBatteryValues(that);
		let chargeLevel = 0;
		let chargingState = 2;
		let lowBattery = false;
		if (!isEmptyObject(batteryValues)) {
			chargeLevel = batteryValues.STORAGE.chargeLevel;
			lowBattery = batteryValues.STORAGE.critical;
			if (batteryValues.STORAGE.status == "Idle") {
				chargingState = 0;
			}
			else {
				batteryValues.connections.forEach(element => {
					if (element.to == "STORAGE"){
	//    		inProgress
						chargingState = 1;
					}
					else {
						chargingState = 0;
					}
				})
			}
		}
		that.battery
			.getCharacteristic(Characteristic.BatteryLevel)
				.updateValue(chargeLevel)
			.getCharacteristic(Characteristic.ChargingState)
				.updateValue(chargingState)
			.getCharacteristic(Characteristic.StatusLowBattery)
				.updateValue(lowBattery)
	}
}


class SolarEdgeInverter {
	constructor(log, config) {
		this.log = log
		this.config = config
		this.current = this.config.current;

		if(this.current) {
			this.currentPower = new Service.LightSensor("Current Power","Current Power");
			this.currentWatts = this.config.currentWatts;
		}

		if(this.config.last_day) {
			this.lastDayPower = new Service.LightSensor("Current Day", "Current Day")
		}

		if(this.config.last_month) {
			this.lastMonth = new Service.LightSensor("Current Month", "Current Month")
		}

		if(this.config.last_year) {
			this.lastYear = new Service.LightSensor("Current Year", "Current Year")
		}

		if(this.config.life_time) {
			this.lifeTime = new Service.LightSensor("Life Time", "Life Time")
		}

		if(this.config.battery) {
			this.battery = new Service.BatteryService("Battery Level", "Battery Level")
		}

		this.name = this.config.name;
		this.manufacturer = this.config.manufacturer || "SolarEdge";
		this.model = this.config.model || "Inverter";
		this.serial = this.config.serial || "solaredge-inverter-1";
		this.siteID = this.config.site_id;
		this.apiKey = this.config.api_key;
		this.update_interval = (this.config.update_interval || 15) * 60 * 1000;
		this.debug = this.config.debug || false;
		this.minLux = this.config.min_lux || DEF_MIN_LUX;
		this.maxLux = this.config.max_lux || DEF_MAX_LUX;

		if(this.currentPower || this.lastDayPower || this.lastMonth || this.lastYear || this.lifeTime || this.battery) {
			update(this);
			setInterval ( async() => {
				if(this.currentPower || this.lastDayPower || this.lastMonth || this.lastYear || this.lifeTime) {
					const accessoryValue = await getAccessoryValue(this);
					let power = 0;
					if (accessoryValue) {
						if(this.currentPower) {
							if(parseFloat(accessoryValue.currentPower.power) > 0) {
								if(this.currentWatts) {
									power = Math.abs(Math.round((accessoryValue.currentPower.power + Number.EPSILON) *10) /10)
								} else {
									power = Math.abs(Math.round(((accessoryValue.currentPower.power / 1000) + Number.EPSILON) *10) /10)
								}
							}
							this.currentPower
								.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
								.updateValue(power ? power : 0.0001)
						}
						if(this.lastDayPower) {
							power = Math.abs(Math.round(((accessoryValue.lastDayData.energy / 1000) + Number.EPSILON) *10) /10)
							this.lastDayPower
								.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
								.updateValue(power ? power : 0.0001)
						}
						if(this.lastMonth) {
							power = Math.abs(Math.round(((accessoryValue.lastMonthData.energy / 1000) + Number.EPSILON) *10) /10)
							this.lastMonth
								.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
								.updateValue(power ? power : 0.0001)
						}
						if(this.lastYear) {
							power = Math.abs(Math.round(((accessoryValue.lastYearData.energy / 1000) + Number.EPSILON) *10) /10)
							this.lastYear
								.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
								.updateValue(power ? power : 0.0001)
						}
						if(this.lifeTime) {
							power = Math.abs(Math.round(((accessoryValue.lifeTimeData.energy / 1000) + Number.EPSILON) *10) /10)
							this.lifeTime
								.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
								.updateValue(power ? power : 0.0001)
						}
					}
				}
				if(this.battery){
					const batteryValues = await getBatteryValues(this);
					let chargeLevel = 0;
					let chargingState = 2;
					let lowBattery = false;
					if (!isEmptyObject(batteryValues)) {
						chargeLevel = batteryValues.STORAGE.chargeLevel;
						lowBattery = batteryValues.STORAGE.critical;
						if (batteryValues.STORAGE.status == "Idle") {
							chargingState = 0;
						}
						else {
							batteryValues.connections.forEach(element => {
								if (element.to == "STORAGE"){
				//    		inProgress
									chargingState = 1;
								}
								else {
									chargingState = 0;
								}
							})
						}
					}
					this.battery
						.getCharacteristic(Characteristic.BatteryLevel)
							.updateValue(chargeLevel)
						.getCharacteristic(Characteristic.ChargingState)
							.updateValue(chargingState)
						.getCharacteristic(Characteristic.StatusLowBattery)
							.updateValue(lowBattery)
				}
			}, this.update_interval)
		}
	}

	getServices () {
		const informationService = new Service.AccessoryInformation()
			.setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
			.setCharacteristic(Characteristic.Model, this.model)
			.setCharacteristic(Characteristic.SerialNumber, this.serial)

		const services = [informationService]

		if(this.current) {
			this.currentPower.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
			services.push(this.currentPower);
		}

		if(this.lastDayPower) {
			this.lastDayPower.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
			services.push(this.lastDayPower);
		}

		if(this.lastMonth) {
			this.lastMonth.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
			services.push(this.lastMonth);
		}

		if(this.lastYear) {
			this.lastYear.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
			services.push(this.lastYear);
		}

		if(this.lifeTime) {
			this.lifeTime.getCharacteristic(Characteristic.CurrentAmbientLightLevel)
			services.push(this.lifeTime);
		}

		if(this.battery) {
			this.battery.getCharacteristic(Characteristic.BatteryLevel)
			this.battery.getCharacteristic(Characteristic.ChargingState)
			this.battery.getCharacteristic(Characteristic.StatusLowBattery)
						services.push(this.battery);
		}

		return services
	}
}
